File: extension/src/score/textNearDuplication.ts
import coefficients from "../../../schema/minhash-coefficients.json";

// mersenne prime, so a*h stays exact
const MODULUS = (1n << 61n) - 1n;

const FNV_OFFSET_BASIS = 14695981039346656037n;
const FNV_PRIME = 1099511628211n;
const MASK_64 = (1n << 64n) - 1n;

export const DEFAULT_SHINGLE_SIZE = 5;
export const DEFAULT_NUM_PERMUTATIONS = 128;
export const DEFAULT_BANDS = 32;
export const DEFAULT_ROWS = 4;
export const DEFAULT_JACCARD_THRESHOLD = 0.7;

export interface ReviewForNearDuplication {
  text: string | null;
}

export interface TextNearDuplicationResult {
  duplicateReviewShare: number | null;
  clusterCount: number;
  largestClusterShare: number;
}

export interface TextNearDuplicationOptions {
  shingleSize?: number;
  numPermutations?: number;
  bands?: number;
  rows?: number;
  jaccardThreshold?: number;
  signatureCache?: WeakMap<ReviewForNearDuplication, bigint[]>;
  linkCache?: DuplicateLinkCache;
}

export interface DuplicateLinks {
  population: object;
  bands: number;
  rows: number;
  threshold: number;
  linked: (readonly bigint[])[];
}

export type DuplicateLinkCache = WeakMap<readonly bigint[], DuplicateLinks>;

export function fnv1a64(input: string): bigint {
  let hash = FNV_OFFSET_BASIS;
  const bytes = new TextEncoder().encode(input);
  for (const byte of bytes) {
    hash ^= BigInt(byte);
    hash = (hash * FNV_PRIME) & MASK_64;
  }
  return hash;
}

export function shingle(text: string, shingleSize: number): Set<string> {
  const normalized = text.toLowerCase().trim().replace(/\s+/g, " ");
  if (normalized.length <= shingleSize) {
    return new Set([normalized]);
  }
  const shingles = new Set<string>();
  for (let i = 0; i <= normalized.length - shingleSize; i++) {
    shingles.add(normalized.slice(i, i + shingleSize));
  }
  return shingles;
}

export function exactJaccard(a: ReadonlySet<string>, b: ReadonlySet<string>): number {
  if (a.size === 0 && b.size === 0) {
    return 1;
  }
  let intersection = 0;
  for (const value of a) {
    if (b.has(value)) {
      intersection++;
    }
  }
  const union = a.size + b.size - intersection;
  return union === 0 ? 0 : intersection / union;
}

export function minhashSignature(
  shingles: ReadonlySet<string>,
  numPermutations: number,
): bigint[] {
  const hashes = Array.from(shingles, (value) => fnv1a64(value) % MODULUS);
  const signature: bigint[] = [];
  for (let i = 0; i < numPermutations; i++) {
    const pair = coefficients[i] as [string, string] | undefined;
    if (pair === undefined) {
      throw new Error(`numPermutations exceeds the ${coefficients.length} shared coefficients`);
    }
    const a = BigInt(pair[0]);
    const b = BigInt(pair[1]);
    let min: bigint | null = null;
    for (const h of hashes) {
      const value = (a * h + b) % MODULUS;
      if (min === null || value < min) {
        min = value;
      }
    }
    signature.push(min ?? 0n);
  }
  return signature;
}

export function estimateJaccard(signatureA: readonly bigint[], signatureB: readonly bigint[]): number {
  let matches = 0;
  for (let i = 0; i < signatureA.length; i++) {
    if (signatureA[i] === signatureB[i]) {
      matches++;
    }
  }
  return matches / signatureA.length;
}

const bandKeyCache = new WeakMap<readonly bigint[], { bands: number; rows: number; keys: string[] }>();

function bandKeys(signature: readonly bigint[], bands: number, rows: number): string[] {
  const cached = bandKeyCache.get(signature);
  if (cached !== undefined && cached.bands === bands && cached.rows === rows) {
    return cached.keys;
  }
  const parts = signature.map((value) => value.toString());
  const keys: string[] = [];
  for (let band = 0; band < bands; band++) {
    const start = band * rows;
    keys.push(`${band}:${parts.slice(start, start + rows).join(",")}`);
  }
  bandKeyCache.set(signature, { bands, rows, keys });
  return keys;
}

class UnionFind {
  private readonly parent: number[];

  constructor(size: number) {
    this.parent = Array.from({ length: size }, (_, i) => i);
  }

  find(x: number): number {
    let root = x;
    while (this.parent[root] !== root) {
      root = this.parent[root] as number;
    }
    let cur = x;
    while (this.parent[cur] !== root) {
      const next = this.parent[cur] as number;
      this.parent[cur] = root;
      cur = next;
    }
    return root;
  }

  union(x: number, y: number): void {
    const rootX = this.find(x);
    const rootY = this.find(y);
    if (rootX !== rootY) {
      this.parent[rootX] = rootY;
    }
  }
}

export function textNearDuplication(
  reviews: readonly ReviewForNearDuplication[],
  options: TextNearDuplicationOptions = {},
): TextNearDuplicationResult {
  const shingleSize = options.shingleSize ?? DEFAULT_SHINGLE_SIZE;
  const numPermutations = options.numPermutations ?? DEFAULT_NUM_PERMUTATIONS;
  const bands = options.bands ?? DEFAULT_BANDS;
  const rows = options.rows ?? DEFAULT_ROWS;
  const threshold = options.jaccardThreshold ?? DEFAULT_JACCARD_THRESHOLD;

  const seeded = (review: ReviewForNearDuplication): bigint[] | undefined => {
    const cached = options.signatureCache?.get(review);
    return cached !== undefined && cached.length === numPermutations ? cached : undefined;
  };

  const byText = new Map<string, bigint[]>();
  const signatures: (readonly bigint[])[] = [];
  for (const review of reviews) {
    const cached = seeded(review);
    if (cached !== undefined) {
      signatures.push(cached);
      continue;
    }
    if (review.text === null || review.text.length === 0) {
      continue;
    }
    const shared = byText.get(review.text);
    const signature = shared ??
      minhashSignature(shingle(review.text, shingleSize), numPermutations);
    if (shared === undefined) {
      byText.set(review.text, signature);
    }
    options.signatureCache?.set(review, signature);
    signatures.push(signature);
  }
  if (signatures.length < 2) {
    return {
      duplicateReviewShare: signatures.length === 0 ? null : 0,
      clusterCount: 0,
      largestClusterShare: 0,
    };
  }

  const unionFind = new UnionFind(signatures.length);

  const firstIndexOf = new Map<readonly bigint[], number>();
  const distinct: number[] = [];
  for (let i = 0; i < signatures.length; i++) {
    const signature = signatures[i] as readonly bigint[];
    const first = firstIndexOf.get(signature);
    if (first === undefined) {
      firstIndexOf.set(signature, i);
      distinct.push(i);
    } else {
      unionFind.union(i, first);
    }
  }

  const links = options.linkCache ?? new WeakMap<readonly bigint[], DuplicateLinks>();
  if (!linksCover(links, signatures, distinct, bands, rows, threshold)) {
    findLinks(links, signatures, distinct, bands, rows, threshold);
  }

  for (const i of distinct) {
    const entry = links.get(signatures[i] as readonly bigint[]) as DuplicateLinks;
    for (const other of entry.linked) {
      const j = firstIndexOf.get(other);
      if (j !== undefined) {
        unionFind.union(i, j);
      }
    }
  }

  const clusterSizes = new Map<number, number>();
  for (let i = 0; i < signatures.length; i++) {
    const root = unionFind.find(i);
    clusterSizes.set(root, (clusterSizes.get(root) ?? 0) + 1);
  }

  let duplicateReviewCount = 0;
  let clusterCount = 0;
  let largestClusterSize = 0;
  for (const size of clusterSizes.values()) {
    if (size >= 2) {
      duplicateReviewCount += size;
      clusterCount++;
      largestClusterSize = Math.max(largestClusterSize, size);
    }
  }

  return {
    duplicateReviewShare: duplicateReviewCount / signatures.length,
    clusterCount,
    largestClusterShare: largestClusterSize / signatures.length,
  };
}

function linksCover(
  links: DuplicateLinkCache,
  signatures: readonly (readonly bigint[])[],
  distinct: readonly number[],
  bands: number,
  rows: number,
  threshold: number,
): boolean {
  const first = links.get(signatures[distinct[0] as number] as readonly bigint[]);
  if (
    first === undefined || first.bands !== bands || first.rows !== rows ||
    first.threshold !== threshold
  ) {
    return false;
  }
  return distinct.every((i) =>
    links.get(signatures[i] as readonly bigint[])?.population === first.population
  );
}

function findLinks(
  links: DuplicateLinkCache,
  signatures: readonly (readonly bigint[])[],
  distinct: readonly number[],
  bands: number,
  rows: number,
  threshold: number,
): void {
  const buckets = new Map<string, number[]>();
  for (const i of distinct) {
    for (const key of bandKeys(signatures[i] as readonly bigint[], bands, rows)) {
      const bucket = buckets.get(key);
      if (bucket === undefined) {
        buckets.set(key, [i]);
      } else {
        bucket.push(i);
      }
    }
  }

  const width = signatures.length;
  const candidatePairs = new Set<number>();
  for (const bucket of buckets.values()) {
    if (bucket.length < 2) {
      continue;
    }
    for (let i = 0; i < bucket.length; i++) {
      for (let j = i + 1; j < bucket.length; j++) {
        candidatePairs.add((bucket[i] as number) * width + (bucket[j] as number));
      }
    }
  }

  const linked = new Map<number, (readonly bigint[])[]>();
  for (const i of distinct) {
    linked.set(i, []);
  }
  for (const pair of candidatePairs) {
    const left = Math.floor(pair / width);
    const right = pair % width;
    const similarity = estimateJaccard(
      signatures[left] as readonly bigint[],
      signatures[right] as readonly bigint[],
    );
    if (similarity > threshold) {
      linked.get(left)?.push(signatures[right] as readonly bigint[]);
      linked.get(right)?.push(signatures[left] as readonly bigint[]);
    }
  }

  const population = {};
  for (const i of distinct) {
    links.set(signatures[i] as readonly bigint[], {
      population,
      bands,
      rows,
      threshold,
      linked: linked.get(i) as (readonly bigint[])[],
    });
  }
}

File: research/verdict_research/features/text_near_duplication.py
import json
from dataclasses import dataclass
from pathlib import Path

# a mersenne prime, large enough that fnv1a64 hashes reduce into it with
# negligible bias, small enough that a*h stays exact under python's
# arbitrary precision integers either way
MODULUS = (1 << 61) - 1

FNV_OFFSET_BASIS = 14695981039346656037
FNV_PRIME = 1099511628211
MASK_64 = (1 << 64) - 1

DEFAULT_SHINGLE_SIZE = 5
DEFAULT_NUM_PERMUTATIONS = 128
DEFAULT_BANDS = 32
DEFAULT_ROWS = 4
DEFAULT_JACCARD_THRESHOLD = 0.7

_COEFFICIENTS_PATH = Path(__file__).parents[3] / "schema" / "minhash-coefficients.json"
with open(_COEFFICIENTS_PATH) as f:
    # coefficients are stored as decimal strings because they exceed a
    # javascript safe integer, and a plain json import there would round them
    COEFFICIENTS: list[tuple[int, int]] = [(int(a), int(b)) for a, b in json.load(f)]


@dataclass
class ReviewForNearDuplication:
    text: str | None


@dataclass
class TextNearDuplicationResult:
    duplicate_review_share: float | None
    cluster_count: int
    largest_cluster_share: float


def fnv1a64(text: str) -> int:
    hash_value = FNV_OFFSET_BASIS
    for byte in text.encode("utf-8"):
        hash_value ^= byte
        hash_value = (hash_value * FNV_PRIME) & MASK_64
    return hash_value


def shingle(text: str, shingle_size: int) -> set[str]:
    normalized = " ".join(text.lower().split())
    if len(normalized) <= shingle_size:
        return {normalized}
    return {normalized[i : i + shingle_size] for i in range(len(normalized) - shingle_size + 1)}


def exact_jaccard(a: set[str], b: set[str]) -> float:
    if not a and not b:
        return 1.0
    intersection = len(a & b)
    union = len(a) + len(b) - intersection
    return intersection / union if union else 0.0


def minhash_signature(shingles: set[str], num_permutations: int) -> list[int]:
    if num_permutations > len(COEFFICIENTS):
        raise ValueError(f"numPermutations exceeds the {len(COEFFICIENTS)} shared coefficients")
    hashes = [fnv1a64(value) % MODULUS for value in shingles]
    signature = []
    for a, b in COEFFICIENTS[:num_permutations]:
        signature.append(min((a * h + b) % MODULUS for h in hashes))
    return signature


def estimate_jaccard(signature_a: list[int], signature_b: list[int]) -> float:
    matches = sum(1 for x, y in zip(signature_a, signature_b, strict=True) if x == y)
    return matches / len(signature_a)


class _UnionFind:
    def __init__(self, size: int) -> None:
        self.parent = list(range(size))

    def find(self, x: int) -> int:
        root = x
        while self.parent[root] != root:
            root = self.parent[root]
        while self.parent[x] != root:
            self.parent[x], x = root, self.parent[x]
        return root

    def union(self, x: int, y: int) -> None:
        root_x, root_y = self.find(x), self.find(y)
        if root_x != root_y:
            self.parent[root_x] = root_y


# SPEC.md 5.4: minhash with 128 permutations over character 5 grams, banded
# lsh, cluster reviews above 0.7 jaccard similarity. bands and rows and the
# output shape are not specified there; this is a proposal, not a ratified
# spec line.
def text_near_duplication(
    reviews: list[ReviewForNearDuplication],
    shingle_size: int = DEFAULT_SHINGLE_SIZE,
    num_permutations: int = DEFAULT_NUM_PERMUTATIONS,
    bands: int = DEFAULT_BANDS,
    rows: int = DEFAULT_ROWS,
    jaccard_threshold: float = DEFAULT_JACCARD_THRESHOLD,
) -> TextNearDuplicationResult:
    eligible = [review.text for review in reviews if review.text]
    if len(eligible) < 2:
        share = None if len(eligible) == 0 else 0.0
        return TextNearDuplicationResult(
            duplicate_review_share=share, cluster_count=0, largest_cluster_share=0.0
        )

    signatures = [
        minhash_signature(shingle(text, shingle_size), num_permutations) for text in eligible
    ]

    buckets: dict[tuple[int, tuple[int, ...]], list[int]] = {}
    for band in range(bands):
        start = band * rows
        for i, signature in enumerate(signatures):
            key = (band, tuple(signature[start : start + rows]))
            buckets.setdefault(key, []).append(i)

    union_find = _UnionFind(len(signatures))
    candidate_pairs: set[tuple[int, int]] = set()
    for bucket in buckets.values():
        if len(bucket) < 2:
            continue
        for i in range(len(bucket)):
            for j in range(i + 1, len(bucket)):
                candidate_pairs.add((bucket[i], bucket[j]))

    for left, right in candidate_pairs:
        similarity = estimate_jaccard(signatures[left], signatures[right])
        if similarity > jaccard_threshold:
            union_find.union(left, right)

    cluster_sizes: dict[int, int] = {}
    for i in range(len(signatures)):
        root = union_find.find(i)
        cluster_sizes[root] = cluster_sizes.get(root, 0) + 1

    duplicate_review_count = 0
    cluster_count = 0
    largest_cluster_size = 0
    for size in cluster_sizes.values():
        if size >= 2:
            duplicate_review_count += size
            cluster_count += 1
            largest_cluster_size = max(largest_cluster_size, size)

    return TextNearDuplicationResult(
        duplicate_review_share=duplicate_review_count / len(eligible),
        cluster_count=cluster_count,
        largest_cluster_share=largest_cluster_size / len(eligible),
    )
